File: QuickSort.cpp
#include <iostream>

void swap(int* arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

int partition(int* arr, int lo, int hi) {
    int pivot = arr[hi];
    int i = lo;
    for(int j=lo; j<hi; ++j)
        if(arr[j] <= pivot)
            swap(arr, i++, j);
    swap(arr, i, hi);
    return i;
}

void quicksort(int* arr, int lo, int hi) {
    if(lo < hi) {
        int p = partition(arr, lo, hi);
        quicksort(arr, lo, p-1);
        quicksort(arr, p+1, hi);
    }
}

int main() {
    int sample[10] = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    quicksort(sample, 0, 9);

    for(int i=0; i<10; i++)
        std::cout << sample[i] << ",";
}

File: QuickSort.js
var quickSort = function(array) {
    var swap = function(a, b) {
        var buf = array[a];
        array[a] = array[b];
        array[b] = buf;
    };

    var partition = function(p, r) {
        var q = p;
        for(var j = p; j < r; ++j) {
            if(array[j] <= array[r]) {
                swap(j, q++);
            }
        }
        swap(r, q);
        return q;
    };

    var execute = function(p, r) {
        if(p < r) {
            var q = partition(p, r);
            execute(p, q-1);
            execute(q+1, r);
        }
    }

    execute(0, array.length-1);
};


var sample = [3, 7, 8, 5, 2, 1, 9, 5, 4];
quickSort(sample);
console.log(sample);
